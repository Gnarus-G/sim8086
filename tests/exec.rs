use std::process::Command;

use insta::assert_display_snapshot;

macro_rules! test_with {
    ($file:literal) => {
        let _ = Command::new("nasm")
            .arg(format!("./fixtures/exec/{}.asm", $file))
            .status()
            .unwrap();

        let app_output = assert_cmd::Command::cargo_bin("sim8086")
            .unwrap()
            .arg(format!("./fixtures/exec/{}", $file))
            .arg("--exec")
            .output()
            .map(|out| String::from_utf8(out.stdout).unwrap())
            .unwrap();

        insta::with_settings!({ description => $file }, {
            assert_display_snapshot!(app_output);
        })
    };
}

#[test]
fn movs() {
    test_with!("immediate_movs");
    test_with!("register_movs");
}

#[test]
fn add_sub_cmp() {
    test_with!("add_sub_cmp");
}

#[test]
fn ip_tracking() {
    test_with!("ip_register");
}

#[test]
fn conditional_jumps() {
    // test_with!("add_sub_cmp");
}
