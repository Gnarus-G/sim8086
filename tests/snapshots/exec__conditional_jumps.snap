---
source: tests/exec.rs
description: conditional_jumps
expression: app_output
---
mov cx, 3 ; cx:0x0->0x3, ip:0x0->0x3, 
mov bx, 1000 ; bx:0x0->0x3e8, ip:0x3->0x6, 
add bx, word 10 ; bx:0x3e8->0x3f2, ip:0x6->0x9, 
sub cx, word 1 ; cx:0x3->0x2, ip:0x9->0xc, 
jne $-8 ; ip:0xc->0x6, 
add bx, word 10 ; bx:0x3f2->0x3fc, ip:0x6->0x9, 
sub cx, word 1 ; cx:0x2->0x1, ip:0x9->0xc, 
jne $-8 ; ip:0xc->0x6, 
add bx, word 10 ; bx:0x3fc->0x406, ip:0x6->0x9, 
sub cx, word 1 ; cx:0x1->0x0, ip:0x9->0xc, flags:->Z
jne $-8 ; ip:0xc->0xe, 

Final registers:
   ax: 0x0000 (0)
   bx: 0x0406 (1030)
   cx: 0x0000 (0)
   dx: 0x0000 (0)
   sp: 0x0000 (0)
   bp: 0x0000 (0)
   si: 0x0000 (0)
   di: 0x0000 (0)
   ip: 0x000e (14)
flags: Z

